php/libraries/Candidate.class.inc

```diff
@@ -281,14 +266,32 @@ class Candidate
                     );
 
         if ($config->getSetting('useExternalID') === 'true') {
-            $externalID = Candidate::_generateExternalID();
+            $externalID = (new ExternalIDGenerator())->generate();
             $setArray['ExternalID'] = $externalID;
         }
 
-        $db->insert('candidate', $setArray);
+        // Generate a new CandID and insert it into the DB. In case of a race
+        // condition, keep trying new IDs until an unused one is found.
+        $generator = new CandIDGenerator();
+        do {
```

Comment:
This do|while can be delegated to the generate function of CandIDGenerator since it already have a dependency on Database for getExistingIDs.
That way, the Generators will have the responsibility of returning a valid Id.